.github/scripts/rust/lint.rs

```diff
@@ -0,0 +1,255 @@
+        }
+
+        for lint in file.deny {
+            self.0.insert(lint, LintLevel::Deny);
+        }
+
+        for lint in file.forbid {
+            self.0.insert(lint, LintLevel::Forbid);
+        }
+
+        for lint in file.default {
+            self.0.remove(&lint);
+        }
+    }
+
+    pub fn into_file(self) -> LintFile {
+        self.0.into_iter().fold(LintFile::default(), |mut acc, (lint, level)| {
+            match level {
+                LintLevel::Allow => acc.allow.insert(lint),
+                LintLevel::Warn => acc.warn.insert(lint),
+                LintLevel::Deny => acc.deny.insert(lint),
+                LintLevel::Forbid => acc.forbid.insert(lint),
+            };
+
+            acc
+        })
+    }
+}
+
+#[derive(Deserialize, Default)]
+struct LintFile {
+    #[serde(default)]
+    allow: HashSet<String>,
+    #[serde(default)]
+    warn: HashSet<String>,
+    #[serde(default)]
+    deny: HashSet<String>,
+    #[serde(default)]
+    forbid: HashSet<String>,
+    #[serde(default)]
+    default: HashSet<String>,
+}
+
+fn collect_lints(cwd: &Path) -> LintFile {
+    let mut active = cwd;
+    let mut files = vec![];
+
+    // go up the directory tree and find all files in `.config/lints.toml`
+    loop {
+        let candidate = active.join(".config/lints.toml");
+
+        if candidate.exists() && candidate.is_file() {
+            files.push(candidate);
+        }
+
+        // we do this until we reach the root of the git repo (a `.git` folder is present)
+        if active.join(".git").exists() && active.join(".git").is_dir() {
+            break;
+        }
+
+        // ... or we have no parent anymore
+        if let Some(parent) = active.parent() {
+            active = parent;
+        } else {
+            break;
+        }
+    }
+
+    // we now apply them into a single configuration this allows for easy overwrites in child directories
+    files.reverse();
+
+    let mut lints = Lints::new();
+
+    for file in files {
+        let contents = fs::read_to_string(&file).expect(&format!("should be able to read {}", file.display()));
+
+        let file = toml::from_str(&contents).expect(&format!("should be valid toml ({})", file.display()));
+        lints.apply(file);
+    }
+
+    lints.into_file()
+}
+
+fn collect_cargo(cwd: &Path) -> LintFile {
+    let path = cwd.join(".cargo/config.toml");
+
+    if !path.exists() || !path.is_file() {
+        panic!(".cargo/config.toml needs to exist")
+    }
+
+    let contents = fs::read_to_string(&path).expect(&format!("should be able to read {}", path.display()));
+
+    // brute-force, search for the line:
+    // ## START CLIPPY LINTS ##
+    // and then:
+    // ## END CLIPPY LINTS ##
+    // take all values in between
+
+    if !contents.contains("## START CLIPPY LINTS ##") || !contents.contains("## END CLIPPY LINTS ##") {
+        panic!("malformed .cargo/config.toml, please add the required markers")
+    }
```

Comment:
Not necessarily happy about this error message as it does not show how the file is expected to look like

Reply:
reworked in [287f64a](https://github.com/hashintel/hash/pull/2075/commits/287f64abe7ef298b32ea879b6a602d14f160a8f5)